rfd/00xx-pod-access-requests.md

```diff
@@ -0,0 +1,555 @@
+Teleport will only allow access to the Pods included in the list, even for cases
+where other roles allow unrestricted access to other Pods or the Kubernetes RBAC
+allows access to other Pods. `kubernetes_pods` is a subset of the super-set allowed
+by the user's Kubernetes RBAC principals.
+
+To do this, Teleport will intercept requests destined for Pod endpoints - Appendix A -
+to either authorize requests that are destined for specific Pods or to
+filter the response coming from Kubernetes API to include only the list of
+authorized Pods.
+
+Once Teleport intercepts a request destined for a specific Pod, it validates if
+the allowed list includes the Pod name. If the Pod is present, Teleport will
+forward the request to the Kubernetes API server. In cases where the Pod name is
+available but is not included in the allowed list, Teleport will reject the
+request and will not forward it to the Kubernetes API.
+
+If the request is destined to a Pod endpoint that does not have Pod names associated -
+`kubectl get pods`, `kubectl get pods --all-namespaces`,
+Teleport will forward the request to the cluster and will modify the received
+response. Once it receives a response from the Kubernetes API server, Teleport
+will filter out any pod the user should not have access.
+
+2. Teleport will delegate access control to Kubernetes RBAC.
+
+When a user requests access to a specific Pod that currently he has no access to,
+Teleport will modify his Kubernetes RBAC principals to include a new group that grants
+him access to the Pod. This indeed will extend his permissions when interacting
+with the Kubernetes Cluster instead of restricting them as in approach 1.
+
+To do this, Teleport will create temporary `Roles` in the Kubernetes cluster and
+will associate them with Kubernetes groups that the user can impersonate.
+
+Once Teleport receives an access request to a Pod, it will create the Kubernetes
+RBAC `Role` and will create `RoleBinding` to assign it to an RBAC group. Afterward,
+Teleport will include the newly created group in the user's certificate.
+
+Once the user has the certificate with the new group embedded, Teleport will
+forward the certificate details into the Kubernetes cluster as Impersonation Headers.
+
+When the access request expires, Teleport will clean the temporary Kubernetes
+RBAC `Role` and `RoleBinding` objects.
+
+## UX: Request Flow
+
+Extending [RFD 59](https://github.com/gravitational/teleport/blob/master/rfd/0059-search-based-access-requests.md),
+Teleport will consider Kubernetes Pods as native resources.
+This allows users to search Pods that normally they do not
+have access to and request access to them. Besides requesting access to
+specific Pods - using the Pod full name, Teleport allows users to request
+access to Pods that follow a pattern like Pods automatically created from a
+Deployment - `deploymentname-*-*`.
+
+```bash
+$ tsh request search --kind pod --search pod
```

Comment:
I think this flow will rarely be used, I think we should support the flow most people will use most of the time:

```bash
$ kubectl exec
... teleport intercepts this and creates access request (similarly to how tsh ssh works with on demand access requests)
```

Reply:
@klizhentas @r0mant 

I am afraid that to implement this flow, we will need to change the current implementation.

Currently, each time a user creates an access request, `tsh` waits for it to be approved and stores the `request_id` in the user's certificate. Only after `tsh` connects to the ssh node.

The problem is that when `kubectl` calls the exec plugin to generate the certificates for the cluster, it does not provide any information about the resource we are trying to access - in this case, the Pod. Thus `tsh kube credentials` cannot request an access request. On the other hand, `kubectl` can't refresh the certificates - it has no idea about it.

I thought of two ways we could solve this problem:

1. The `kube_service` executes the access request as soon as it identifies that the user doesn't have access and:
-    i) sends the command to wait for admin approval. This command refreshes the certificates if the admin approves the request.
-   ii)  Blocks until the request is approved and disconnects. The next time the user calls `tsh kube credentials`, Auth Server will include the `request_id` in the certificate.

3. Teleport stores the `request_ids` server-side - `kube_proxy/service`, and each time the user requests a Kube endpoint, the service inserts the request_ids associated with the active requests for that user. We would need a way to discard the active access requests.

Neither of these is ideal.

Reply:
@tigrato I suggest we go with a much simpler solution, in addition to the flow you created, we can roll out:

`tsh kubectl exec` command, that will mirror and pass through all arguments, except it will detect if the pod will require special access and create ad-hoc access request before pass-through to `kubectl`. Will this work?

Reply:
@klizhentas 

Yes, this simpler approach will work.

I assume that `tsh kubectl edit pod(s,/, )pod1 pod2 pod3`, and `tsh kubectl port-forward ...` will also have to be supported, right?

I'm a bit wary of using `tsh kubectl` for two reasons:
- tsh would have to catch every command destined for pods - kubectl has a lot of aliases for pods: `po`, `pod`, `pods`, `pod/`.
- For me, it's unnatural to use the tsh prefix before kubectl. We can alias `kubectl="tsh kubectl"` but who would do it?

One possible solution to avoid so many possible combinations is to create a proxy and intercept the `kubectl` HTTP requests destined for pods when `tsh kubectl` is called.

Reply:
@klizhentas Just wanted to follow up on this ^^^ - how do you feel about the approach with having a local Kubernetes proxy `tsh proxy kube` similar to what we have for database/application access? With this approach it would actually be the proxy that would be requesting access to pods e.g. `tsh proxy kube nginx --namespace=teleport`. A couple nice properties of this approach that I see are:

- We actually have customers asking for a local Kubernetes proxy with ALB support similar to what we did for database access so that would lay a groundwork for that. I think it was actually even considered as this quarter's stretch goal.
- This should integrate nicely with Teleport Connect as it similarly manages local proxies for databases.

So users would start the local proxy (themselves or with Connect) and be able to use `kubectl` to connect through it. We can still keep `tsh kubectl exec` as a single-command convenience wrapper that would start local proxy on demand and make kubectl connect through it (same approach as with `tsh db connect`).

WDYT?